bundles/org.openhab.binding.vdr/src/main/java/org/openhab/binding/vdr/internal/svdrp/SVDRPClientImpl.java

```diff
@@ -0,0 +1,409 @@
+import java.util.regex.Pattern;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link SVDRPClientImpl} encapsulates all calls to the SVDRP interface of a VDR
+ *
+ * @author Matthias Klocke - Initial contribution
+ */
+@NonNullByDefault
+public class SVDRPClientImpl implements SVDRPClient {
+
+    private final Logger logger = LoggerFactory.getLogger(SVDRPClientImpl.class);
+
+    private String host;
+    private int port = 6419;
+    private String charset = "UTF-8";
+    private String version = "";
+
+    private final String welcomeMessagePattern = "([0-9]{3})([ -])(.*)";
+    private Pattern patternMessage = Pattern.compile(welcomeMessagePattern);
+
+    private static int timeout = 3000;
+
+    private @Nullable Socket socket = null;
+    private @Nullable BufferedWriter out = null;
+    private @Nullable BufferedReader in = null;
+
+    public SVDRPClientImpl(String host, int port) {
+        super();
+        this.host = host;
+        this.port = port;
+    }
+
+    public SVDRPClientImpl(String host, int port, String charset) {
+        super();
+        this.host = host;
+        this.port = port;
+        this.charset = charset;
+    }
+
+    @Override
+    protected void finalize() {
+        try {
+            this.closeConnection();
+        } catch (Exception ex) {
+            logger.trace("Closing SVDRPConnection not successful: {}", ex.getMessage());
+        }
+    }
+
+    /**
+     *
+     * Open VDR Socket Connection
+     *
+     * @throws IOException if an IO Error occurs
+     */
+    @SuppressWarnings("null")
+    @Override
+    public void openConnection() throws SVDRPConnectionException, SVDRPParseResponseException {
+        if (socket == null || socket.isClosed()) {
+            socket = new Socket();
+        }
+        try {
+            InetSocketAddress isa = new InetSocketAddress(host, port);
+            socket.connect(isa, timeout);
+            socket.setSoTimeout(timeout);
+
+            out = new BufferedWriter(new OutputStreamWriter(socket.getOutputStream(), charset), 8192);
+            in = new BufferedReader(new InputStreamReader(socket.getInputStream(), charset), 8192);
+
+            // read welcome message and init version & charset
+            SVDRPResponse res = null;
+
+            res = execute(null);
+
+            if (res.getCode() == 220) {
+                SVDRPWelcome welcome = SVDRPWelcome.parse(res.getMessage());
+                this.charset = welcome.getCharset();
+                this.version = welcome.getVersion();
+            } else {
+                throw new SVDRPParseResponseException(res);
+            }
+        } catch (IOException e1) {
+            try {
+                // cleanup after timeout
+                out.close();
+                in.close();
+
+                socket.close();
+            } catch (Exception e) {
+            }
```

Comment:
You could move these in the `finally` block to close these also if an unchecked exception (e.g. RuntimeException) is raised, which you don't catch here (which is ok).

Reply:
I made it a checked exception as I only want to close the resources in the error case. Therefore I didn't put in into finally block.